File: src/test/java/org/example/kalkulationsprogramm/repository/AnfrageRepositorySearchTest.java
package org.example.kalkulationsprogramm.repository;

import org.example.kalkulationsprogramm.domain.Anfrage;
import org.example.kalkulationsprogramm.domain.Kunde;
import org.example.kalkulationsprogramm.repository.KundeRepository;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;

import java.time.LocalDate;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

@DataJpaTest
class AnfrageRepositorySearchTest {

        @Autowired
        private AnfrageRepository anfrageRepository;
        @Autowired
        private KundeRepository kundeRepository;

        @Test
        void search_respects_umlauts_and_ascii_separately() {
                Anfrage a1 = new Anfrage();
                Kunde k1 = new Kunde();
                k1.setName("Haentschel");
                k1.setKundennummer("K1001");
                kundeRepository.saveAndFlush(k1);
                a1.setKunde(k1);
                a1.setBauvorhaben("Testbau");
                a1.setAnlegedatum(LocalDate.now());
                anfrageRepository.saveAndFlush(a1);

                Anfrage a2 = new Anfrage();
                Kunde k2 = new Kunde();
                k2.setName("H\u00E4ntschel"); // Häntschel
                k2.setKundennummer("K1002");
                kundeRepository.saveAndFlush(k2);
                a2.setKunde(k2);
                a2.setBauvorhaben("Testbau 2");
                a2.setAnlegedatum(LocalDate.now());
                anfrageRepository.saveAndFlush(a2);

                // Query: Häntschel should match only Häntschel
                List<Anfrage> umlaut = anfrageRepository.search("H\u00E4ntschel", null, null, null, null);
                assertThat(umlaut).extracting(a -> a.getKunde().getName())
                                .containsExactlyInAnyOrder("H\u00E4ntschel");

                // Query: Haentschel should match only Haentschel
                List<Anfrage> ascii = anfrageRepository.search("Haentschel", null, null, null, null);
                assertThat(ascii).extracting(a -> a.getKunde().getName())
                                .containsExactlyInAnyOrder("Haentschel");
        }

        @Test
        void search_bauvorhaben_does_not_cross_map_umlauts() {
                Anfrage a1 = new Anfrage();
                Kunde k1 = new Kunde();
                k1.setName("Kunde");
                k1.setKundennummer("K1003");
                kundeRepository.saveAndFlush(k1);
                a1.setKunde(k1);
                a1.setBauvorhaben("M\u00FCllerstra\u00DFe Dach"); // Müllerstraße Dach
                a1.setAnlegedatum(LocalDate.now());
                anfrageRepository.saveAndFlush(a1);

                Anfrage a2 = new Anfrage();
                Kunde k2 = new Kunde();
                k2.setName("Kunde");
                k2.setKundennummer("K1004");
                kundeRepository.saveAndFlush(k2);
                a2.setKunde(k2);
                a2.setBauvorhaben("Muellerbau");
                a2.setAnlegedatum(LocalDate.now());
                anfrageRepository.saveAndFlush(a2);

                // Query: Mueller should match only 'Muellerbau'
                List<Anfrage> res1 = anfrageRepository.search(null, "Mueller", null, null, null);
                assertThat(res1).extracting(Anfrage::getBauvorhaben)
                                .containsExactlyInAnyOrder("Muellerbau");

                // Query: Müller should match only the umlaut variant
                List<Anfrage> res2 = anfrageRepository.search(null, "M\u00FCller", null, null, null);
                assertThat(res2).extracting(Anfrage::getBauvorhaben)
                                .containsExactlyInAnyOrder("M\u00FCllerstra\u00DFe Dach");
        }
}

File: src/test/kotlin/org/example/kalkulationsprogramm/repository/AnfrageRepositorySearchTest.kt
package org.example.kalkulationsprogramm.repository

import org.assertj.core.api.Assertions.assertThat
import org.example.kalkulationsprogramm.domain.Anfrage
import org.example.kalkulationsprogramm.domain.Kunde
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import java.time.LocalDate

@DataJpaTest
class AnfrageRepositorySearchTest {

    @Autowired
    private lateinit var anfrageRepository: AnfrageRepository

    @Autowired
    private lateinit var kundeRepository: KundeRepository

    @Test
    fun searchRespectsUmlautsAndAsciiSeparately() {
        val anfrage1 = Anfrage()
        val kunde1 = Kunde().apply {
            name = "Haentschel"
            kundennummer = "K1001"
        }
        kundeRepository.saveAndFlush(kunde1)
        anfrage1.kunde = kunde1
        anfrage1.bauvorhaben = "Testbau"
        anfrage1.anlegedatum = LocalDate.now()
        anfrageRepository.saveAndFlush(anfrage1)

        val anfrage2 = Anfrage()
        val kunde2 = Kunde().apply {
            name = "Häntschel"
            kundennummer = "K1002"
        }
        kundeRepository.saveAndFlush(kunde2)
        anfrage2.kunde = kunde2
        anfrage2.bauvorhaben = "Testbau 2"
        anfrage2.anlegedatum = LocalDate.now()
        anfrageRepository.saveAndFlush(anfrage2)

        val umlaut = anfrageRepository.search("Häntschel", null, null, null, null)
        assertThat(umlaut).extracting<String> { it.kunde!!.name }
            .containsExactlyInAnyOrder("Häntschel")

        val ascii = anfrageRepository.search("Haentschel", null, null, null, null)
        assertThat(ascii).extracting<String> { it.kunde!!.name }
            .containsExactlyInAnyOrder("Haentschel")
    }

    @Test
    fun searchBauvorhabenDoesNotCrossMapUmlauts() {
        val anfrage1 = Anfrage()
        val kunde1 = Kunde().apply {
            name = "Kunde"
            kundennummer = "K1003"
        }
        kundeRepository.saveAndFlush(kunde1)
        anfrage1.kunde = kunde1
        anfrage1.bauvorhaben = "Müllerstraße Dach"
        anfrage1.anlegedatum = LocalDate.now()
        anfrageRepository.saveAndFlush(anfrage1)

        val anfrage2 = Anfrage()
        val kunde2 = Kunde().apply {
            name = "Kunde"
            kundennummer = "K1004"
        }
        kundeRepository.saveAndFlush(kunde2)
        anfrage2.kunde = kunde2
        anfrage2.bauvorhaben = "Muellerbau"
        anfrage2.anlegedatum = LocalDate.now()
        anfrageRepository.saveAndFlush(anfrage2)

        val res1 = anfrageRepository.search(null, "Mueller", null, null, null)
        assertThat(res1).extracting<String> { it.bauvorhaben }
            .containsExactlyInAnyOrder("Muellerbau")

        val res2 = anfrageRepository.search(null, "Müller", null, null, null)
        assertThat(res2).extracting<String> { it.bauvorhaben }
            .containsExactlyInAnyOrder("Müllerstraße Dach")
    }
}
